### PR TITLE
common/HeartbeatMap: print stack info of unhealth thread.

### PR DESCRIPTION
**Pull request by majianpeng**
We use heartbeatmap to detect thread whether health. If timeout over
suicide of thread, we assert osd. But now we assert the check-thread of
heartbeatmap rather than unhealth thread.
We hope dump stack info of unhealth thread to find usefull info.

Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

### CODE REVIEW

**Comment by liewegas** [comment]
Instead of (sometimes) passing the thread id to add_worker, instead we could set the id from the reset function (if it hasn't been set yet). That way all users are captured after first use..

File: src/common/HeartbeatMap.cc [hunk 1]
```diff
@@ -78,7 +79,10 @@ bool HeartbeatMap::_check(const heartbeat_handle_d *h, const char *who, time_t n
   if (was && was < now) {
     ldout(m_cct, 1) << who << " '" << h->name << "'"
 		    << " had suicide timed out after " << h->suicide_grace << dendl;
-    assert(0 == "hit suicide timeout");
+    if (h->thread_id == 0)
+      assert(0 == "hit suicide timeout");
+    else
+      pthread_kill(h->thread_id, SIGABRT);
```
Comment (liewegas) [inline src/common/HeartbeatMap.cc]: Does this trigger a stack trace dump on the other thread?  SIGABRT or something?

**Comment by liewegas** [comment]
@majianpeng ping

**Comment by majianpeng** [comment]
@liewegas . For the reset function, i don't know. Could you explain in detail? 
I tested on osd, SIGABRT can work. 
But in global_init()
>>if (g_conf->fatal_signal_handlers)
>>    install_standard_sighandlers();
It depend on g_conf->fatal_signal_handlers. 
Have you better method? Thanks!

**Comment by liewegas** [comment]
@majianpeng Let's change the pthread_kill branch to also sleep(1) and then  still assert.  That way we will always die, even if the kill fails for some reason.

File: src/common/HeartbeatMap.h [hunk 2]
```diff
@@ -42,19 +42,20 @@ namespace ceph {
...
-    : name(n), grace(0), suicide_grace(0)
+    : name(n), thread_id(0), grace(0), suicide_grace(0)
   { }
 };
 
 class HeartbeatMap {
  public:
   // register/unregister
-  heartbeat_handle_d *add_worker(const std::string& name);
+  heartbeat_handle_d *add_worker(const std::string& name, pthread_t thread_id = 0);
```
Comment (athanatos) [inline src/common/HeartbeatMap.h]: No need for an optional param here since all callers appear to supply it?

**Comment by liewegas** [comment]
Also, please rebase

**Comment by majianpeng** [comment]
update. please review!

**Comment by liewegas** [comment]
compile error:
<pre>
test/heartbeat_map.cc: In member function ‘virtual void HeartbeatMap_Healthy_Test::TestBody()’:
test/heartbeat_map.cc:25:46: error: no matching function for call to ‘ceph::HeartbeatMap::add_worker(const char [4])’
   heartbeat_handle_d *h = hm.add_worker("one");
                                              ^
test/heartbeat_map.cc:25:46: note: candidate is:
In file included from test/heartbeat_map.cc:16:0:
./common/HeartbeatMap.h:58:23: note: ceph::heartbeat_handle_d* ceph::HeartbeatMap::add_worker(const string&, pthread_t)
   heartbeat_handle_d *add_worker(const std::string& name, pthread_t thread_id);
                       ^
./common/HeartbeatMap.h:58:23: note:   candidate expects 2 arguments, 1 provided
test/heartbeat_map.cc: In member function ‘virtual void HeartbeatMap_Unhealth_Test::TestBody()’:
test/heartbeat_map.cc:36:46: error: no matching function for call to ‘ceph::HeartbeatMap::add_worker(const char [4])’
   heartbeat_handle_d *h = hm.add_worker("one");
                                              ^
test/heartbeat_map.cc:36:46: note: candidate is:
In file included from test/heartbeat_map.cc:16:0:
./common/HeartbeatMap.h:58:23: note: ceph::heartbeat_handle_d* ceph::HeartbeatMap::add_worker(const string&, pthread_t)
   heartbeat_handle_d *add_worker(const std::string& name, pthread_t thread_id);
                       ^
./common/HeartbeatMap.h:58:23: note:   candidate expects 2 arguments, 1 provided
</pre>

**Pull request merged**